File: LeetCode/Python Solutions/Reorder List/reorder.py
class ListNode:
    def __init__(self, val=0, next=None):
<<<<<<<<<<<<<<  ✨ Codeium Command 🌟 >>>>>>>>>>>>>>>>
-        self.val = val
-        self.next = next

class Solution:
+    """
+    @param head: The head of the linked list
+    @return: nothing
+    """
<<<<<<<  2794d946-bfee-4795-b6aa-864d746a96a3  >>>>>>>
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return
        
        # Find the middle of the list
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # Reverse the second half
        prev, curr = None, slow.next
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        slow.next = None  # Cut off the first half
        
        # Merge two halves
        first, second = head, prev
        while second:
            temp1, temp2 = first.next, second.next
            first.next = second
            second.next = temp1
            first, second = temp1, temp2

File: LeetCode/Java Solutions/Reorder List/reorder.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
   public void reorderList(ListNode head) {
       if (head == null || head.next == null) return;
       
       // Find the middle of the list
       ListNode slow = head, fast = head;
       while (fast != null && fast.next != null) {
           slow = slow.next;
           fast = fast.next.next;
       }
       
       // Reverse the second half of the list
       ListNode prev = null, curr = slow.next;
       while (curr != null) {
           ListNode nextTemp = curr.next;
           curr.next = prev;
           prev = curr;
           curr = nextTemp;
       }
       slow.next = null; // Split the list into two parts
       
       // Merge two halves
       ListNode first = head, second = prev;
       while (second != null) {
           ListNode temp1 = first.next, temp2 = second.next;
           first.next = second;
           second.next = temp1;
           first = temp1;
           second = temp2;
       }
   }
}
